toolbox/relnotes/main.go

```diff
@@ -0,0 +1,606 @@
+	f.WriteString("```\n")
+
+	log.Printf("CI job status fetched.")
+	return nil
+}
+
+// createBody creates the general documentation, example and downloads table body for the
+// markdown file.
+func createBody(c *github.Client, f *os.File, releaseTag, branch, docURL, exampleURL, releaseTars string) {
+	var title string
+	if *preview {
+		title = "Branch "
+	}
+
+	if releaseTag == "HEAD" || releaseTag == branchHead {
+		title += branch
+	} else {
+		title += releaseTag
+	}
+
+	if *preview {
+		f.WriteString("**Release Note Preview - generated on " + time.Now().Format("Mon Jan  2 15:04:05 MST 2006") + "**\n")
+	}
+
+	f.WriteString("\n# " + title + "\n\n")
+	f.WriteString(fmt.Sprintf("[Documentation](%s) & [Examples](%s)\n\n", docURL, exampleURL))
+
+	if releaseTars != "" {
+		f.WriteString("## Downloads for " + title + "\n\n")
+		createDownloadsTable(f, releaseTag, "", releaseTars+"/kubernetes.tar.gz", releaseTars+"/kubernetes-src.tar.gz")
+		createDownloadsTable(f, releaseTag, "Client Binaries", releaseTars+"/kubernetes-client*.tar.gz")
+		createDownloadsTable(f, releaseTag, "Server Binaries", releaseTars+"/kubernetes-server*.tar.gz")
+		createDownloadsTable(f, releaseTag, "Node Binaries", releaseTars+"/kubernetes-node*.tar.gz")
+		f.WriteString("\n")
+	}
+}
+
+// createDownloadTable creates table of download link and sha256 hash for given file.
+func createDownloadsTable(f *os.File, releaseTag, heading string, filename ...string) {
+	var urlPrefix string
+
+	if *releaseBucket == "kubernetes-release" {
+		urlPrefix = "https://dl.k8s.io"
+	} else {
+		urlPrefix = "https://storage.googleapis.com/" + *releaseBucket + "/release"
+	}
+
+	if heading != "" {
+		f.WriteString(fmt.Sprintf("\n### %s\n", heading))
+	}
+
+	f.WriteString("\n")
+	f.WriteString("filename | sha256 hash\n")
+	f.WriteString("-------- | -----------\n")
+
+	files := make([]string, 0)
+	for _, name := range filename {
+		fs, _ := filepath.Glob(name)
+		for _, v := range fs {
+			files = append(files, v)
+		}
+	}
+
+	for _, file := range files {
+		fn := extractFileName(file)
+		sha, _ := u.GetSha256(file)
+		f.WriteString(fmt.Sprintf("[%s](%s/%s/%s) | `%s`\n", fn, urlPrefix, releaseTag, fn, sha))
+	}
+}
+
+// extractFileName takes a string and returns the file name after last '/'.
+func extractFileName(filePath string) string {
+	i := strings.LastIndex(filePath, "/")
+	if i != -1 {
+		return filePath[i+1 : len(filePath)]
+	}
+	return filePath
+}
+
+// minorReleases performs a minor (vX.Y.0) release by fetching the release template and aggregate
+// previous release in series.
+func minorRelease(f *os.File, release, draftURL, changelogURL string) {
+
+	// Check for draft and use it if available
+	log.Printf("Checking if draft release notes exist for %s...", release)
+
+	resp, err := http.Get(draftURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	// TODO: find a better way to tell failed response
+	if err == nil && (resp.StatusCode == 200 || resp.StatusCode == 304) {
```

Comment:
Why we consider 304 as a success?

Reply:
@yutongz I saw some times the file was cached and the status code was 304 when I was testing, but maybe that was on the web browser not the program. Not really sure about this one...

Reply:
In that case, make sure when you decide it failed, print out the err message or status code. Maybe we can find out some other magic. Actually, you may always make sure your program don't eat err message.

Reply:
Status code 304 should not happen here, since we are doing an unconditional GET. I think it's fine to fail if it's anything other than 200.

Reply:
@enisoc Done.